hydra-kafka10/src/main/java/com/addthis/hydra/kafka/bundle/KafkaSource.java

```diff
@@ -0,0 +1,283 @@
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class KafkaSource extends TaskDataSource {
+
+    private static final Logger log = LoggerFactory.getLogger(KafkaSource.class);
+
+    @JsonProperty(required = true)
+    private String zookeeper;
+    @JsonProperty(required = true)
+    private String topic;
+    @JsonProperty
+    private String initialOffset = "earliest";
+    @JsonProperty
+    private String dateFormat = "YYMMdd";
+    @JsonProperty
+    private String markDir = "marks";
+    /** Ignore the mark directory */
+    @JsonProperty
+    private boolean ignoreMarkDir;
+    /** Specifies conversion to bundles.  If null, then uses DataChannelCodec.decodeBundle */
+    @JsonProperty
+    private BundleizerFactory format;
+
+    @JsonProperty
+    private int fetchThreads = 1;
+    @JsonProperty
+    private int decodeThreads = 1;
+    @JsonProperty
+    private int queueSize = 10000;
+    @JsonProperty
+    private int seedBrokers = 3;
+
+    @JsonProperty
+    private TaskRunConfig config;
+    @JsonProperty
+    private int pollRetries = 180;
+
+    @JsonProperty
+    private String bootstrapServers;
+    @JsonProperty
+    private Map<String, String> overrideProperties = new HashMap();
+
+    Properties consumerProperties;
+    PageDB<SimpleMark> markDb;
+    AtomicBoolean running;
+    LinkedBlockingQueue<BundleWrapper> bundleQueue;
+    final ConcurrentMap<String, Long> sourceOffsets = new ConcurrentHashMap<>();
+
+    private ExecutorService fetchExecutor;
+    private ExecutorService decodeExecutor;
+
+    @Override
+    public Bundle next() {
+        if (!running.get()) {
+            return null;
+        }
+        BundleWrapper bundle;
+        try {
+            bundle = bundleQueue.poll(pollRetries, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            // reset interrupt status
+            Thread.currentThread().interrupt();
+            // fine to leave bundles on the queue
+            return null;
+        }
+        if (bundle == null) {
+            throw new DataChannelError(
+                    "giving up on kafka source next() after waiting: " + pollRetries + " seconds");
+        }
+        if (bundle == bundleQueueEndMarker) {
+            // add back end-marker in case someone continues calling peek/next on the source
+            bundleQueue.add(bundleQueueEndMarker);
+            return null;
+        }
+        // Although the bundle queue is guaranteed to be ordered by offsets, concurrent calls to
+        // next() can complete (and update offsets) in arbitrary order.  Assuming the threads calling
+        // next() will not be interrupted, we can just always take the max offset per partition since all
+        // earlier bundles will eventually be returned to some next() call.
+        sourceOffsets.compute(bundle.sourceIdentifier, (partition, previous) ->
+                previous == null ? bundle.offset : Math.max(previous, bundle.offset));
```

Comment:
Im not sure, but I think the offset should be max(previous, offset) + 1.  Otherwise, when the source runs a second time, I think it would re-consume the last message.
Should run a test where we kick a job twice to see if duplicates the last message in the topic.